ratis-client/src/main/java/org/apache/ratis/client/api/DataStreamApi.java

```diff
@@ -44,4 +48,7 @@ default DataStreamOutput stream() {
 
   /** Create a stream by providing a customized header message. */
   DataStreamOutput stream(ByteBuffer headerMessage);
+
+  /** Create a stream by providing a customized header message and route table. */
+  DataStreamOutput stream(ByteBuffer headerMessage, Map<RaftPeerId, List<RaftPeerId>> routingTable);
```

Comment:
@szetszwo RoutingTable has been defined as a interface as follows, maybe we need to defined another class ?
```
 interface RoutingTable {
    List<RaftPeerId> getSuccessors(RaftPeerId peer);
  }
```

Reply:
Let's add a Builder.
```
public interface RoutingTable {
  Set<RaftPeerId> getSuccessors(RaftPeerId peerId);

  class Builder {
    private final Map<RaftPeerId, Set<RaftPeerId>> map = new HashMap<>();

    private Set<RaftPeerId> computeIfAbsent(RaftPeerId peerId) {
      return map.computeIfAbsent(peerId, key -> new HashSet<>());
    }

    public Builder addSuccessor(RaftPeerId peerId, RaftPeerId successor) {
      computeIfAbsent(peerId).add(successor);
      return this;
    }

    public Builder addSuccessors(RaftPeerId peerId, List<RaftPeerId> successors) {
      computeIfAbsent(peerId).addAll(successors);
      return this;
    }

    public Builder addSuccessors(RaftPeerId peerId, RaftPeerId... successors) {
      return addSuccessors(peerId, Arrays.asList(successors));
    }

    public RoutingTable build() {
      return peerId -> Optional.ofNullable(map.get(peerId)).orElseGet(Collections::emptySet);
    }
  }
}
```
In the code, we should keeping using RoutingTable instead of Map.  It will give us more flexibility.

Btw, it should return a Set instead of List.  My bad!

Reply:
@szetszwo Maybe we also need to expose the map to user, because we need encode map to proto, the lambda `peerId -> Optional.ofNullable(map.get(peerId)).orElseGet(Collections::emptySet)` is not enough.

Reply:
That is a good point.  We should add a toProto() method to the interface.
```
  public interface RoutingTable {
    Set<RaftPeerId> getSuccessors(RaftPeerId peerId);

    RoutingTableProto toProto();
  }
```

Reply:
@szetszwo Sorry, not familar with Java, but the following code compiled error.
```
public interface RoutingTable {
  Set<RaftPeerId> getSuccessors(RaftPeerId peerId);

  RoutingTableProto toProto();

  class Builder {
    private final Map<RaftPeerId, Set<RaftPeerId>> map = new HashMap<>();

    private Set<RaftPeerId> computeIfAbsent(RaftPeerId peerId) {
      return map.computeIfAbsent(peerId, key -> new HashSet<>());
    }

    public Builder addSuccessor(RaftPeerId peerId, RaftPeerId successor) {
      computeIfAbsent(peerId).add(successor);
      return this;
    }

    public Builder addSuccessors(RaftPeerId peerId, List<RaftPeerId> successors) {
      computeIfAbsent(peerId).addAll(successors);
      return this;
    }

    public Builder addSuccessors(RaftPeerId peerId, RaftPeerId... successors) {
      return addSuccessors(peerId, Arrays.asList(successors));
    }

    public RoutingTable build() {
      return peerId -> Optional.ofNullable(map.get(peerId)).orElseGet(Collections::emptySet);
    }
  }
}
```

Reply:
The build() method needs to be updated.  We should also change the Builder to use a reference so that it cannot be built more than once.
```
public interface RoutingTable {
  Set<RaftPeerId> getSuccessors(RaftPeerId peerId);

  RoutingTableProto toProto();

  class Builder {
    private final AtomicReference<Map<RaftPeerId, Set<RaftPeerId>>> ref = new AtomicReference<>(new HashMap<>());

    private Set<RaftPeerId> computeIfAbsent(RaftPeerId peerId) {
      return Optional.ofNullable(ref.get())
          .map(map -> map.computeIfAbsent(peerId, key -> new HashSet<>()))
          .orElseThrow(() -> new IllegalStateException("Already built"));
    }

    public Builder addSuccessor(RaftPeerId peerId, RaftPeerId successor) {
      computeIfAbsent(peerId).add(successor);
      return this;
    }

    public Builder addSuccessors(RaftPeerId peerId, Collection<RaftPeerId> successors) {
      computeIfAbsent(peerId).addAll(successors);
      return this;
    }

    public Builder addSuccessors(RaftPeerId peerId, RaftPeerId... successors) {
      return addSuccessors(peerId, Arrays.asList(successors));
    }

    public RoutingTable build() {
      return Optional.ofNullable(ref.getAndSet(null))
          .map(RoutingTable::newRoutingTable)
          .orElseThrow(() -> new IllegalStateException("Already built"));
    }
  }

  static RoutingTable newRoutingTable(Map<RaftPeerId, Set<RaftPeerId>> map){
    return new RoutingTable() {
      @Override
      public Set<RaftPeerId> getSuccessors(RaftPeerId peerId) {
        return Optional.ofNullable(map.get(peerId)).orElseGet(Collections::emptySet);
      }

      @Override
      public RoutingTableProto toProto() {
        return RoutingTableProto.newBuilder().addAllRoutingTable(ProtoUtils.toRouteProtos(map)).build();
      }
    };
  }
}
```

Reply:
@szetszwo Thanks, really learn a lot.